src/main/java/input_feed/distributed/TextToDistributedR1CS.java

```diff
@@ -0,0 +1,220 @@
+                    final long startIndex = part * (totalSize / numExecutors);
+                    final long partSize = part == numExecutors ?
+                            totalSize % (totalSize / numExecutors) : totalSize / numExecutors;
+
+                    final ArrayList<Tuple2<Long, FieldT>> assignment = new ArrayList<>();
+                    for (long i = startIndex; i < startIndex + partSize; i++) {
+                        assignment.add(new Tuple2<>(i, serialAssignment.get((int) i)));
+                    }
+                    return assignment.iterator();
+                }).persist(this.config().storageLevel());
+
+        final Assignment<FieldT> primary = new Assignment<>(serialAssignment.subList(0, numInputs));
+
+        return new Tuple2<>(primary, distributedAssignment);
+    }
+
+    private JavaPairRDD<Long, LinearTerm<FieldT>>
+    distributedCombination(
+            String fileName,
+            ArrayList<Integer> partitions,
+            int numConstraints,
+            boolean negate
+    ){
+        final int numPartitions = this.config().numPartitions();
+
+        // Require at least one constraint per partition!
+        assert(numConstraints >= numPartitions);
+
+        JavaPairRDD<Long, LinearTerm<FieldT>> result;
+        try {
+
+            final BufferedReader br = new BufferedReader(new FileReader(fileName));
+            // TODO - this is not the best for performance
+            Map<Integer, LinearCombination<FieldT>> constraintMap = serializeReader(br, negate);
+            br.close();
+
+            result = this.config().sparkContext().parallelize(partitions, numPartitions).flatMapToPair(part -> {
+                final long partSize;
+                if (part == numPartitions && numConstraints < 2 * numPartitions) {
+                    partSize = numConstraints - numPartitions;
+                } else {
+                    partSize = part == numPartitions ? numConstraints %
+                            (numConstraints / numPartitions) : numConstraints / numPartitions;
+                }
+
+                final ArrayList<Tuple2<Long, LinearTerm<FieldT>>> T = new ArrayList<>();
+                for (long i = 0; i < partSize; i++) {
+
+                    final long index = part * (numConstraints / numPartitions) + i;
+                    if (constraintMap.containsKey((int) index)){
+                        LinearCombination<FieldT> currRow = constraintMap.get((int) index);
+                        for (LinearTerm term: currRow.terms()) {
+                            T.add(new Tuple2<>(index, term));
+                        }
+                    }
+                }
+                return T.iterator();
+            });
+
+            return result;
+        } catch (Exception e) {
+            System.err.println("Error: " + e.getMessage());
+        }
+        return null;
+    }
+
+    public Map<Integer, LinearCombination<FieldT>> serializeReader(BufferedReader br, boolean negate) {
```

Comment:
I wonder if it would be more efficient to ship the filename and an offset to the hosts, instead of serializing the reader (which is essentially loading the file in memory)

Reply:
That could be a good idea. I would need to know how to attach the volume to all the slave nodes probably rework the way that the data that is fed into each component. This would require a better understanding of how this spark actually handles these partitioned datasets.